src/Spice86.Core/Emulator/VM/CpuSpeedLimit/CpuCycleThrottler.cs

```diff
@@ -0,0 +1,107 @@
+﻿namespace Spice86.Core.Emulator.VM.CpuSpeedLimit;
+
+using Spice86.Core.Emulator.CPU;
+using static Spice86.Core.Emulator.Devices.Timer.Timer;
+
+using System.Diagnostics;
+
+/// <summary>
+/// Throttles CPU execution to maintain a target number of cycles per millisecond,
+/// using a budget-based approach similar to DOSBox.
+/// </summary>
+public class CpuCycleThrottler : CycleLimiterBase {
+    // Keep track of timing and cycles
+    private readonly SpinWait _spinner = new();
+    private readonly Stopwatch _stopwatch = new();
+    private long _cycleRemain;
+    private long _lastTicks;
+
+    // Constants for cycle control
+    private const int MaxCyclesPerWindow = 20;
+    private const int CyclesUp = 1000;
+    private const int CyclesDown = 1000;
+    private const int MaxCyclesPerMs = 60000;
+    private const int MinCyclesPerMs = 100;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CpuCycleThrottler"/> class.
+    /// </summary>
+    /// <param name="targetCpuCyclesPerMs">The target CPU cycles per millisecond.</param>
+    public CpuCycleThrottler(int targetCpuCyclesPerMs) {
+        if (targetCpuCyclesPerMs > 0) {
+            TargetCpuCyclesPerMs = Math.Clamp(
+            targetCpuCyclesPerMs,
+            MinCyclesPerMs,
+            MaxCyclesPerMs);
+        } else {
+            TargetCpuCyclesPerMs = Math.Clamp(
+            ICyclesLimiter.RealModeCpuCyclesPerMs,
+            MinCyclesPerMs,
+            MaxCyclesPerMs);
+        }
+
+        _stopwatch.Start();
+        _lastTicks = _stopwatch.ElapsedTicks;
+    }
+
+    /// <inheritdoc/>
+    internal override void RegulateCycles(State cpuState) {
+        if (!cpuState.IsRunning) {
+            return;
+        }
+
+        // If we still have cycle budget left, decrement and return
+        if (_cycleRemain > 0) {
+            _cycleRemain--;
```

Comment:
couldnt we avoid doing that and compare instead the cpuState.Cycles property with a _targetCyclesForNextMs var?

Reply:
I don't think so ?